5.drug-dependency/scripts/4.drug_latent_dim_correlation.py

```diff
@@ -0,0 +1,263 @@
+    reordered_df = dependency_df[original_feature_names]
+    # Transform models into pca space
+    latent_df = pd.DataFrame(
+        model.transform(reordered_df)
+    )
+
+    # Recode column space and add back model IDs
+    latent_df.columns = [f"z_{x}" for x in range(0, latent_df.shape[1])]
+    latent_df = pd.concat([dependency_df.loc[:, "ModelID"], latent_df], axis="columns")
+    
+    return latent_df
+
+
+def perform_correlation(latent_df, drug_df, model_name, num_components):
+    """
+    Perform Pearson correlation between latent dimensions and drug dependency scores.
+    
+    Parameters:
+    latent_df (pd.DataFrame): Dataframe containing latent dimensions (e.g., PCA/ICA/NMF).
+    drug_df (pd.DataFrame): Dataframe containing drug dependency scores with `ModelID`.
+    model_name (str): Name of the model used to extract the latent dimensions (PCA, ICA, NMF).
+    num_components (int): Number of latent dimensions/components in the model.
+    
+    Returns:
+    pd.DataFrame: Correlation results between latent dimensions and drug scores.
+    """
+    correlation_results = []
+    print("latent_df columns:", latent_df.columns)
+    print("drug_df columns:", drug_df.columns)
+    latent_df.set_index('ModelID', inplace=True)
+
+    # Align both dataframes based on the ModelID
+    common_model_ids = latent_df.index.intersection(drug_df.index)
+    # Filter both dataframes to keep only common ModelIDs
+    latent_df_filtered = latent_df.loc[common_model_ids]
+    prism_df_filtered = drug_df.loc[common_model_ids]
+    # Check the variance of each latent dimension and drug response column
+    latent_variance = latent_df_filtered.var()
+    prism_variance = prism_df_filtered.var()
+
+    # Filter out constant columns (variance == 0)
+    latent_df_filtered = latent_df_filtered.loc[:, latent_variance != 0]
+    prism_df_filtered = prism_df_filtered.loc[:, prism_variance != 0]
+
+    # Loop over each latent dimension and calculate correlation with each drug
+    for latent_col in latent_df_filtered.columns:
+        for drug_col in prism_df_filtered.columns:
+            latent_values = latent_df_filtered[latent_col]
+            drug_values = prism_df_filtered[drug_col]
+
+            # Check if either column is constant
+            if latent_values.nunique() <= 1 or drug_values.nunique() <= 1:
+                corr = np.nan
+            else:
+                # Drop missing values for both columns
+                valid_data = pd.concat([latent_values, drug_values], axis=1).dropna()
+                latent_values_valid = valid_data[latent_col]
+                drug_values_valid = valid_data[drug_col]
+
+                if len(latent_values_valid) > 1 and len(drug_values_valid) > 1:
+                    # Calculate Pearson correlation
+                    corr, p_value = pearsonr(latent_values_valid, drug_values_valid)
+                else:
+                    corr = np.nan
+                    print("nan")
```

Comment:
This function seems like a perfect candidate to set up logging for! This will help clean up your std out and save the output status of the analyses.

Reply:
Consider adding logging in a future PR?

Reply:
Added!